프로젝트 매니져 || [STEP1-1] RED, Taeangel

Taeangel opened the pull request:
@Monsteel 💦🦕💦
토니 STEP3-1 PR 보냅니다!

## 구현한 내용
- [x] realm 적용
- [x] MVVM 구조 수정

## 고민한점 및 조언을 받고 싶은 점

## realm primaryKey 
realm Model 객체를 만들었는데, primaryKey 를 꼭 설정해야 한다
고 해서 
```swift
@Persisted(primaryKey: true) var id: UUID
```
이렇게 key 값을 설정해 주었더니 
Realm() 을 불러오는 과정에서 부터 오류가 나더라구요. 
결국에 false로 값을 바꿔서 key 값을 설정해 주지 않았더니, 문제가 해결 되었습니다.
realm 공식문서에서는 유니크한 key 값을 무조건 설정하라고 하는데, 오류가 나는 원인을 못찾았습니다. 
uuid 를 사용하고 있기 때문에 값이 중복되서 들어갈 염려는 없을 거라고 생각합니다. 
그래도 공식문서에 나온 지침을 따르고 싶은데 조언 부탁드립니다. 

## realm 과 data 의 sync 맞추기 
realm 을 운용하는 service 객체를 만들었고, 
ViewModel 들이 데이터를 가지고 있게 되었는데요,
이 데이터들과 realm 의 데이터의 sync 를 맞춰주기 위해서 모든 행동이 끝날 때마다 
```swift
todoList = todoService.read()
```
ViewModel 데이터에 realm 데이터를 주입해주는 로직을 추가해야 했습니다. 

ViewModel에 데이터를 직접 넣는 방법 보다는 이 방법이 맞다고 생각이 들어서 이렇게 해주었습니다. 

그런데 지금은 수동으로 sync 를 맞춰주고 있다고 생각이 됩니다. 
자동으로 realm 이 바뀌면 ViewModel data 에 업데이트 할 수 있는 방법이 궁금합니다.


## ViewModel 이 View 의 상태를 들고 있어야 하는 점

토니가 저번에 말씀하신 것처럼 View의 상태값을 ViewModel이 알고 있어야 한다고 하셨는데 그렇게 하려면 모든 상태값을 가지는 뷰는 ViewModel을 가지고 있어야 한다고 생각하는데 이 저희가 생각한게 맞을까요?

또 저희가 시도해본 바로는 ViewModel에 state를 가지고 View에 Biding으로 값을 넣어주었더니 Biding 값을 추적해보았을 때, Binding 값을 바꿨음에도 값 자체가 바뀌지 않는 현상이 일어났습니다. 😭😭

cherrishRed committed:
feat: realm 적용

Taeangel committed:
feat: realm CRUD 연습

Taeangel committed:
feat: Realm CRUD 수정

cherrishRed committed:
refactor: TodoListView 에서 TodoService 삭제

cherrishRed committed:
fix: update, create 되지 않는 로직 수정

Taeangel committed:
refactor: ViewModel, Model 코드 정리

Taeangel committed:
refactor: private 접근제어자 수정

Monsteel commented on ProjectManager/ProjectManager/View/AppViewSubviews/TodoList.swift:
```diff
@@ -77,7 +73,7 @@ struct ListCellView: View {
         EditView(
           todo: Todo(id: todo.id, title: todo.title, content: todo.content, date: todo.date, status: todo.status),
```
오류가 발생합니다. ⚠️

생성자를 만들지 않고 변수를 초기화하고, 접근 제어자를 조정하면 이런 휴먼에러가 발생할 가능성이 큽니다.
저는 이러한 이유에서 생성자를 명시적으로 만들어 주는걸 선호합니다.

cherrishRed replied:
```swift
  init(todo: Todo, isShow: Binding<Bool>, viewModel: EditViewModel) {
    self._isShow = isShow
    self.viewModel = viewModel
    self.todo = todo
  }
```
init 을 명시적으로 만들어 주었습니다. 

그런데, init 내부에서 property 를 할당해 주는 순서에 따라서 오류가 나는 일도 발생하나요?

이렇게 순서 대로 init 을 해준것 뿐인데 오류가 나더라구요...
<img width="742" alt="스크린샷 2022-07-21 오후 5 37 02" src="https://user-images.githubusercontent.com/63499481/180169545-cc4bb334-a902-4a03-913a-1754827f1937.png">

순서를 바꿔주니 해결은 되었습니다.

Monsteel reviewed (changes requested):
안녕하세요 red taeangel 이번 스텝도 수고많으셨습니다.

### realm primaryKey

우선, 이 문제는 primaryKey 설정에 오류가 있기 보다는, table의 속성이 바뀌어 발생한 문제인데요

이 문제를 정확하게 이해하기 위해서는 database에 대한 이해도가 필요합니다.

데이터가 채워져있는 상태에서 그 data를 담고있는 table의 속성이 바뀌어서 발생하는 이슈에요.
이런 문제를 해결하기 위해서 migration이라는 개념이 등장하는데, realm에도 동일한 개념이 있습니다.

한번 찾아보시면 좋을 것 같아요.

제 말을 간증하는 현상이 하나 더 있을텐데요, 
`@Persisted(primaryKey: true)`로 설정하고, 앱을 삭제했다가 다시 설치(저장된 데이터를 삭제하고 table을 다시 구축하는 과정)해보시면 문제가 발생하지 않음을 확인하실 수 있을거에요.

### realm 과 data 의 sync 맞추기
[React to Changes(변경에 반응하기)](https://www.mongodb.com/docs/realm/sdk/swift/react-to-changes/) 에서 다루고 있는 내용이 두분이 찾고계신 내용이 맞을까요?

변경되었을 때 호출되는 Closures에서 두분이 적어주신 todoList를 채우는 작업을 할 수 있겠네요.
 
### ViewModel 이 View 의 상태를 들고 있어야 하는 점

> 토니가 저번에 말씀하신 것처럼 View의 상태값을 ViewModel이 알고 있어야 한다고 하셨는데 그렇게 하려면 모든 상태값을 가지는 뷰는 ViewModel을 가지고 있어야 한다고 생각하는데 이 저희가 생각한게 맞을까요?

이론적으로는 맞습니다. 다만, 모든 view에 viewModel을 만드는 행위는 부담이 될 수 있지요.
오버엔지니어링에 대한 고민도 될거구요.

어떤 VIew에는 ViewModel을 만들 것인지, 어떤 뷰에는 안 만들 것인지 명확한 기준을 세워야 합니다.
그렇지 않으면 규칙이 없어 혼란한 상황이 될거에요.

> 또 저희가 시도해본 바로는 ViewModel에 state를 가지고 View에 Biding으로 값을 넣어주었더니 Binding 값을 추적해보았을 때, Binding 값을 바꿨음에도 값 자체가 바뀌지 않는 현상이 일어났습니다. 😭😭

제가 말씀드린 개념적 관점에서의 `State`와, property wrapper `@State`는 다른 개념입니다.

개념적 관점에서의 `State`는 말그대로 화면이 가지는 "상태"를 의미했고,
property wrapper `@State`는 View에서 값 변경에 따라 View가 React하게 작동할 수 있도록 하는 property wrapper를 의미합니다.

ViewModel이 모든 상태를 가져야 한다 에서 "상태"는 개념적 관점에서의 `State`를 의미합니다.

ViewModel은 ObservableObject이기 때문에, 
가지고 있는 상태를 `@Published` property wrapper를 사용해서 가지고 있어야 하겠지요.

이 내용에 대해 혹시 오해가 있으신가 해서 설명을 드렸는데요, 제가 말씀드린 내용과 다른 결의 이슈라면 추가로 말씀 부탁드립니다.

구조적인 내용의 리뷰는 이곳에 남기지 않았습니다.

앞서 디스코드로 말씀드렸듯 오늘 라이브코딩으로 함께

* AppView / AppViewModel
* TodoList / TodoListViewModel

파일들을 다루어 보면서 구조적인 리뷰를 진행합니다.

텍스트로 남길 수 있지만, 제대로된 전달이 힘들것 같아서 디스코드 리뷰를 진행하려고 합니다.
혹시나 텍스트 리뷰를 원하신다면 말씀해주세요. 텍스트로 남겨보겠습니다.

cherrishRed committed:
fix: EditView init 오류 수정

cherrishRed committed:
refactor: realm migration 추가

Taeangel committed:
refactor: AppView, AppViewModel 수정

cherrishRed committed:
refactor: CreateView, CreateViewModel 수정

cherrishRed committed:
refactor: TodoListView, ListViewModel 수정

Taeangel committed:
refactor: ListViewModel 구조변경

cherrishRed committed:
fix: create이후 업데이트 되지 않는 오류 수정

cherrishRed committed:
fix: update 되지않는 로직 수정

cherrishRed committed:
refactor: cell ViewModel 생성

- 수정은 완벽하게 됨
- move 되었을 때 셀이 refresh 되지 않음
- ListViewModel 내부에 ListCellViewModel 이 존재하는 형태로 변경 됨.
- ListCellViewModel 내부에 EditViewModel 이 존재함.
- 문제는 List CellViewModel 이 생기면서 move 가 일어났을 때 update 를 진행하기 어려움.

Taeangel committed:
refactor: editViewModel 프로퍼티 수정

Taeangel committed:
refactor: status이동시 updata되지 않던 에러 처리

Taeangel committed:
refactor: View와 ViewModel 파일정리

Taeangel committed:
refactor: 파일 정리

cherrishRed committed:
fix: edit시 View 가 Update 되지 않는 로직 수정

Taeangel committed:
refactor: 전체적인 네이밍, 컨밴션, 로직 수정

cherrishRed commented:
@Monsteel
토니 저희 MVVM 구조적인 리펙토링 완료 했습니다.

View 와 ViewModel 각각 1:1 대응을 해주었습니다. 
토니가 디스코드에서 조언해주셨던 것 처럼 
View에 있던 상태를 관리하던 프로퍼티들을 모두 ViewModel 로 옮겨 로직적인 부분은 ViewModel에서 처리할 수 있도록 로직을 수정했습니다.

Monsteel reviewed (approved):
두분 모두 수고많으셨습니다~! ☀️
